### ISSUE TITLE
Training is not reproducible

### ISSUE THREAD

**Issue by yermandy** [ultralytics/ultralytics#5276]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

Hello, I did not find any issue regarding non-reproducible training. When I try to train any network, I am getting different loss values at the end of training for the same running script, for instance: 

``` python
import torch

from ultralytics.models.yolo.detect import DetectionTrainer

args = dict(
    model="yolov8n.yaml",
    data="coco-det-person.yaml",
    project="debug",
    epochs=1,
    imgsz=640,
    name="det",
    workers=8,
    batch=128,
    device=0,
    cache="ram",
    optimizer="AdamW",
    lr0=0.001,
)

trainer = DetectionTrainer(overrides=args)

trainer.train()
```

I assumed that the root of the problem lies in random augmentations where we use `random` and `np.random`. When I disabled augmentations, I started getting the same loss for the same script. I expect that workers load batches in not ordered manner causing interruptions across augmentations. Any idea how to fix this behaviour? 

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by yermandy** [comment]
Another point that confirms my observation is that setting workers to 0 also makes loss reproducible

**Comment by yermandy** [comment]
Ok, I found the solution. The question is, do authors want to have it reproducible?

https://discuss.pytorch.org/t/large-differences-in-training-runs-depending-on-worker-number/43352/4

**Comment by yermandy** [comment]
The idea is that in each epoch, each sample gets its own seed for sampling. When the epoch ends, each sample gets a new seed

**Comment by glenn-jocher** [comment]
@yermandy thanks for your question and the research you've done! 

The ability to train models in a reproducible manner is highly desirable to achieve consistent and reliable results. However, ensuring reproducibility can be challenging given the randomness that exists in training processes such as data augmentation, initial weights setting, and data loading.

Your hypothesis about data augmentations seems quite plausible. When using multi-worker data loading, the order of data loading and operations applied on them like data augmentation can have certain variations which might lead to small, but noticeable differences in training loss and consequently, results over runs. 

If you're keen on achieving strict reproducibility, then you can set the number of workers to 0 as that will ensure one consistent order for data loading, and possibly turn off the random data augmentation. However, please keep in mind that this would result in a slower data loading process which might hamper the overall training speed.

Your point on setting seeds at the beginning of each epoch is indeed a potentially useful strategy to ensure each data sample has a consistent "random" state across different epochs. This might limit the effects of workers loading data out of order and still preserve the randomness from the augmentations. This however needs to be implemented with caution and evaluated thoroughly.

We appreciate your willingness to contribute towards this issue. Please feel free to submit a PR and we can further discuss the best ways to handle this. We're eagerly looking forward to your contribution!

**Comment by yermandy** [comment]
Okay, this is interesting. It seems that this behavior appears only when `cache="ram"`. When I do not cache anything to RAM, it seems to work ok

**Comment by yermandy** [comment]
Because caching to RAM is not essential as some datasets are too big to fit into RAM, I am closing this issue with a note that **if you want reproducible results, do not use `cache="ram"` argument**.

**Comment by yermandy** [comment]
I also was not able to identify why exactly this option caused loss to differ in reproducing the run...

**Comment by glenn-jocher** [comment]
@yermandy hello, thank you for bringing this issue to our attention. 

Caching data to RAM can potentially introduce variation due to multi-threading in the data loading process. The caching process can sometimes result in slightly different data access order each time the script is run. This variation in data order, combined with random augmentations, might lead to non-identical mini-batches across different runs, hence causing the variability in reproduced results. 

If reproducibility of training runs is a priority for your use case, we suggest you refrain from using the `cache="ram"` option. Although this might result in slower data loading times, it would ensure consistent access order to data, thereby aiding in reproducibility. 

We will explore measures to ensure consistent data loading order even when `cache="ram"` is used. Feel free to contribute to the repository or check back for updates. Thank you for your interest in Ultralytics YOLOv8.
